### PR TITLE
Add warning when not using max bits for encryption

### PR DESCRIPTION
**Pull request by WilliamBZA**
Connects to https://github.com/Particular/NServiceBus/issues/3313

Checks what the maximum keylength is from the SymmetricAlgorithm class and logs a warning if the key length used is less than the maximum.

This is intended to allow the user to see that they could improve their security.

ping @Particular/nservicebus-maintainers

### CODE REVIEW

File: src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs [hunk 1]
```diff
@@ -198,6 +199,13 @@ static bool IsValidKey(byte[] key)
             using (var rijndael = new RijndaelManaged())
             {
                 var bitLength = key.Length * 8;
+
+                var maxValidKeyBitLength = rijndael.LegalKeySizes.OrderByDescending(keyLength => keyLength.MaxSize).Select(keyLength => keyLength.MaxSize).First();
+                if (bitLength < maxValidKeyBitLength)
+                {
+                    Log.WarnFormat("Encryption key is is {0} bits which is less than the maximum allowed {1} bits. Increasing the key length to {1} bits would provide stronger security.", bitLength, maxValidKeyBitLength);
```
Comment (MarcinHoppe) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: "is" is repeated twice in the warning.

Comment (WilliamBZA) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: Fixed, thanks

Comment (MarcinHoppe) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: I would also re-phrase the second sentence to something like: "Consider using a {1}-bit encryption key to obtain the maximum cipher strength".

Security is multi-faceted and increasing algorithm strength might not automatically provide users with stronger protection.

**Comment by SimonCropp** [comment]
looks good to me

File: src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs (hunk 1 above)
Comment (WilliamBZA) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: Yeah, agreed. Updated.

File: src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs [hunk 2]
```diff
@@ -198,6 +199,13 @@ static bool IsValidKey(byte[] key)
             using (var rijndael = new RijndaelManaged())
             {
                 var bitLength = key.Length * 8;
+
+                var maxValidKeyBitLength = rijndael.LegalKeySizes.OrderByDescending(keyLength => keyLength.MaxSize).Select(keyLength => keyLength.MaxSize).First();
```
Comment (timbussmann) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: why not using `Max(keyLength => keyLength.MaxSize)`?

Comment (WilliamBZA) [inline src/NServiceBus.Core/Encryption/RijndaelEncryptionService.cs]: Because that's what happens when you don't look at your stuff properly.

Updated

**Comment by danielmarbach** [comment]
@timbussmann based on the history in this PR it looks like this has been merged although it didn't compile

image

**Comment by timbussmann** [comment]
@danielmarbach Because of TC build queue the build was queued after it has been merged & deleted.  The changes build & pass all tests.

**Pull request merged**